### PR TITLE
docs: say that all versions in https://apt.llvm.org are available

### PR DESCRIPTION
**Pull request by jnooree**

### CODE REVIEW

**Comment by 2bndy5** [comment]
1. We don't test with versions older than 9.
2. It is preferable that only the major version number is specified. For instance, if `clang-format-12 --version` shows using v12.0.1, then the major version number, `12`, is enough. We don't have exhaustive parsing of semantic version numbers builtin (see [src here](https://github.com/cpp-linter/cpp-linter/blob/977eb162ece45389571fcb8a3bb9b97bca31169d/cpp_linter/__init__.py#L165)). If the version specified is not an integer (major version number), then the version input is interpreted as a path to the binary executables. 

   This point actually proves that the clang-tools-pip and cpp-linter packages are not in complete alignment about dealing with the input `version` string.

File: README.md [hunk 1]
```diff
@@ -82,7 +82,7 @@ jobs:
 
 #### `version`
 
-- **Description**: The desired version of the [clang-tools](https://github.com/cpp-linter/clang-tools-pip) to use. Accepted options are strings which can be 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4 or 3.9.
+- **Description**: The desired version of the [clang-tools](https://github.com/cpp-linter/clang-tools-pip) to use. Accepted options are strings which can be 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4 or 3.9, plus all versions supported by the llvm PPA archives (<https://apt.llvm.org>).
```
[the remark this answers is not in the document]
Comment (2bndy5) [reply README.md:85]: > llvm PPA archives

This is specific to Linux and may imply to some users that only ubuntu runners are supported. We actually do support more than just Linux, and installing all version for other OSs (windows and mac) does not actually employ the LLVM PPA archive.

Therefore, I don't want these changes. @jnooree In the future, please open an issue to discuss any proposed changes before opening a Pull Request (especially one without a description).

[the remark this answers is not in the document]
Comment (jnooree) [reply README.md:85]: Sorry for the empty description; I thought the title was clear enough. Still it is good to mention the newer versions are available (at least for the ubuntu runners) IMO.

Based on your comments, adding version information per platform would be OK? I assume that other platforms are supported by clang-tools-pip and ubuntu is supported by the PPA archive.

**Comment by jnooree** [comment]
> 1. We don't test with versions older than 9.

Should I remove the older ones, or state that they are available but untested?

> 2. It is preferable that only the major version number is specified. For instance, if `clang-format-12 --version` shows using v12.0.1, then the major version number, `12`, is enough. We don't have exhaustive parsing of semantic version numbers builtin (see [src here](https://github.com/cpp-linter/cpp-linter/blob/977eb162ece45389571fcb8a3bb9b97bca31169d/cpp_linter/__init__.py#L165)). If the version specified is not an integer (major version number), then the version input is interpreted as a path to the binary executables.
>    This point actually proves that the clang-tools-pip and cpp-linter packages are not in complete alignment about dealing with the input `version` string.

I'll add the behavior to the docs shortly.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jnooree) [reply README.md:85]: Also, I think the README should mention that this workflow uses clang-tools-pip for fetching the binaries and point to the project for the supported versions. Updating two repository on every new release seems fragile to me.

[the remark this answers is not in the document]
Comment (2bndy5) [reply README.md:85]: > Also, I think the README should mention that this workflow uses clang-tools-pip for fetching the binaries and point to the project for the supported versions. 

This seems like a valid idea. 👍🏼 

> Updating two repository on every new release seems fragile to me.

Its actually good practice to separate different tasks to different software. We automate dependency updates for cpp-linter-action with dependabot. We often like to test cpp-linter-action before publishing a release, especially with updates to cpp-linter or clang-tools-pip.

> Based on your comments, adding version information per platform would be OK?

I'm not sure it is really needed as the cpp-linter clang-tools-pip packages handle the discrepancies between platforms installed binaries.

> I assume that other platforms are supported by clang-tools-pip and ubuntu is supported by the PPA archive.

Yes, for the most part. The LLVM PPA archive doesn't have binaries for releases published before they moved the project to github (only v7+ I think). So, clang-tools-pip can take care of older versions on ubuntu runners also.

**Comment by 2bndy5** [comment]
> Should I remove the older ones, or state that they are available but untested?

Last I checked, the CLI was compatible, so there shouldn't be any concern. In fact we haven't seen anyone use untested versions to far. Thus, this idea seems superfluous.

Maybe we should change "3.9" to "3" since only the major version number is respected and specifying "3.9" might be interpreted as a path. However, I'm more inclined to add better version parsing to the cpp-linter package; so it would handle a full version spec (ie "12.0.1") and correctly only use the major version since clang-tools-pip would have already ensured the specified version is installed.

**Comment by 2bndy5** [comment]
> 2. It is preferable that only the major version number is specified. For instance, if `clang-format-12 --version` shows using v12.0.1, then the major version number, `12`, is enough. We don't have exhaustive parsing of semantic version numbers builtin (see [src here](https://github.com/cpp-linter/cpp-linter/blob/977eb162ece45389571fcb8a3bb9b97bca31169d/cpp_linter/__init__.py#L165)). If the version specified is not an integer (major version number), then the version input is interpreted as a path to the binary executables.
>
>    This point actually proves that the clang-tools-pip and cpp-linter packages are not in complete alignment about dealing with the input `version` string.

No need to document this. https://github.com/cpp-linter/cpp-linter/pull/46 aims to resolve this in code.

**Pull request closed without merging**